File: lib/clusterbuster/pod_files/files.pl
#!/usr/bin/perl

use POSIX;
use strict;
use Fcntl qw(:DEFAULT O_DIRECT);
use JSON;
require "clientlib.pl";

my ($dirs, $files_per_dir, $blocksize, $block_count, $processes, $direct, $drop_cache_service, $drop_cache_port, @dirs) = parse_command_line(@ARGV);

$SIG{TERM} = sub { POSIX::_exit(0); };
my ($bufalign) = 8192;

initialize_timing();

if ($#dirs < 0) {
    @dirs=("/tmp");
}

sub remdir($$) {
    my ($dirname, $oktofail) = @_;
    if (! rmdir("$dirname")) {
	if ($oktofail) {
	    system("rm", "-rf", "$dirname");
	} else {
	    rmdir("$dirname") || die("Can't remove directory $dirname: $!\n");
	}
    }
}

sub makethem($) {
    my ($process) = @_;
    my ($ops) = 0;
    my ($buffer);
    my ($buffer) = 'a' x ($blocksize + $bufalign);
    my ($offset) = unpack('J', pack('p', $buffer)) % $bufalign;
    $offset = $bufalign - $offset;
    my ($fileargs) = O_WRONLY|O_CREAT | ($direct ? O_DIRECT : 0);
    my ($container) = container();
    foreach my $bdir (@dirs) {
	my ($pdir)="$bdir/p$process";
	mkdir("$pdir") || die("Can't create directory $pdir: $!\n");
	$ops++;
	my ($dir)="$pdir/$container";
	mkdir("$dir") || die("Can't create directory $dir: $!\n");
	$ops++;
	foreach my $subdir (0..$dirs-1) {
	    my ($dirname) = "$dir/d$subdir";
	    mkdir("$dirname") || die("Can't create directory $dirname: $!\n");
	    $ops++;
	    foreach my $file (0..$files_per_dir-1) {
		my ($filename) = "$dirname/f$file";
		sysopen(FILE, $filename, $fileargs, 0666) || die "Can't create file $filename: $!\n";
		$ops++;
		foreach my $block (0..$block_count - 1) {
		    if ((my $answer = syswrite(FILE, $buffer, $blocksize, $offset)) != $blocksize) {
			die "Write to $filename failed: $answer $!\n";
		    }
		    $ops++;
		}
		close FILE;
	    }
	}
    }
    return $ops;
}

sub readthem($;$) {
    my ($process, $oktofail) = @_;
    my ($ops) = 0;
    my ($buffer) = 'a' x ($blocksize + $bufalign);
    my ($offset) = unpack('J', pack('p', $buffer)) % $bufalign;
    $offset = $bufalign - $offset;
    my ($fileargs) = O_RDONLY | ($direct ? O_DIRECT : 0);
    my ($container) = container();
    foreach my $bdir (@dirs) {
	my ($pdir)="$bdir/p$process";
	next if ($oktofail && ! -d $pdir);
	my ($dir)="$pdir/$container";
	next if ($oktofail && ! -d $dir);
	foreach my $subdir (0..$dirs-1) {
	    my ($dirname) = "$dir/d$subdir";
	    next if ($oktofail && ! -d $dirname);
	    foreach my $file (0..$files_per_dir-1) {
		my ($filename) = "$dirname/f$file";
		next if ($oktofail && ! -f $filename);
		sysopen(FILE, $filename, $fileargs) || die "Can't open file $filename: $!\n";
		$ops++;
		foreach my $block (0..$block_count - 1) {
		    if ((my $answer = sysread(FILE, $buffer, $blocksize, $offset)) != $blocksize) {
			die "Read from $filename failed: $answer $!\n";
		    }
		    $ops++;
		}
		close(FILE);
	    }
	}
    }
    return $ops;
}

sub removethem($;$) {
    my ($process, $oktofail) = @_;
    my ($ops) = 0;
    my ($container) = container();
    foreach my $bdir (@dirs) {
	my ($pdir)="$bdir/p$process";
	next if ($oktofail && ! -d $pdir);
	my ($dir)="$pdir/$container";
	next if ($oktofail && ! -d $dir);
	foreach my $subdir (0..$dirs-1) {
	    my ($dirname) = "$dir/d$subdir";
	    next if ($oktofail && ! -d $dirname);
	    foreach my $file (0..$files_per_dir-1) {
		my ($filename) = "$dirname/f$file";
		next if ($oktofail && ! -f $filename);
		unlink($filename) || die "Can't remove $filename: $!\n";
		$ops++;
	    }
	    remdir($dirname, $oktofail);
	    $ops++;
	}
	remdir("$dir", $oktofail);
	$ops++;
	remdir("$pdir", $oktofail);
	$ops++;
    }
    return $ops;
}

sub run_one_operation($$$$$$) {
    my ($op_name0, $op_name1, $op_name2, $op_func, $process, $data_start_time) = @_;

    sync_to_controller(idname($process, "start $op_name2"));
    timestamp("$op_name0 files...");
    drop_cache($drop_cache_service, $drop_cache_port);
    my ($ucpu, $scpu) = cputimes();
    my ($op_start_time) = xtime() - $data_start_time;
    my ($ops) = &$op_func($process);
    my ($op_end_time_0) = xtime() - $data_start_time;
    drop_cache($drop_cache_service, $drop_cache_port);
    my ($op_end_time) = xtime() - $data_start_time;
    my ($op_elapsed_time) = $op_end_time - $op_start_time;
    my ($op_elapsed_time_0) = $op_end_time_0 - $op_start_time;
    my ($ucpu, $scpu) = cputimes($ucpu, $scpu);
    my (%answer) = (
	'operation_elapsed_time' => $op_elapsed_time,
	'user_cpu_time' => $ucpu,
	'system_cpu_time' => $scpu,
	'cpu_time' => $ucpu + $scpu,
	'cpu_utilization' => ($ucpu + $scpu) / $op_elapsed_time,
	'operation_start' => $op_start_time,
	'operation_end' => $op_end_time,
	'operations' => $ops,
	'operations_per_second' => $ops / $op_elapsed_time
	);
    if ($op_name2 eq 'read') {
	$answer{'total_files'} = $files_per_dir * $dirs * scalar @dirs;
	$answer{'block_count'} = int $block_count;
	$answer{'block_size'} = int $blocksize;
	$answer{'data_size'} = $blocksize * $block_count * $answer{'total_files'};
	$answer{'data_rate'} = $answer{'data_size'} / $op_elapsed_time_0;
    }
    timestamp("$op_name1 files...");
    sync_to_controller(idname($process, "end $op_name2"));
    return (\%answer, $op_start_time, $op_end_time, $ucpu, $scpu);
}

sub runit($) {
    my ($process) = @_;
    my ($iterations) = 1;
    my ($data_start_time) = xtime();
    # Make sure everything is cleared out first...but don't count the time here.
    removethem($process, 1);
    system("sync");
    my (%extras);

    my ($answer_create, $create_start_time, $create_end_time, $create_ucpu, $create_scpu) =
	run_one_operation('Creating', 'Created', 'create', \&makethem, $process, $data_start_time);
    $extras{'create'} = $answer_create;
    my ($create_et) = $create_end_time - $create_start_time;

    timestamp("Sleeping for 60 seconds");
    sleep(60);
    timestamp("Back from sleep");
    my ($answer_read, $read_start_time, $read_end_time, $read_ucpu, $read_scpu) =
	run_one_operation('Reading', 'Read', 'read', \&readthem, $process, $data_start_time);
    $extras{'read'} = $answer_read;

    timestamp("Sleeping for 60 seconds");
    sleep(60);
    timestamp("Back from sleep");
    my ($answer_remove, $remove_start_time, $remove_end_time, $remove_ucpu, $remove_scpu) =
	run_one_operation('Creating', 'Removed', 'remove', \&removethem, $process, $data_start_time);
    $extras{'remove'} = $answer_remove;
    my ($remove_et) = $remove_end_time - $remove_start_time;
    my ($data_start_time) = $create_start_time;
    my ($data_end_time) = $remove_end_time;
    my ($data_elapsed_time) = $create_et + $remove_et;
    my ($user_cpu) = $create_ucpu + $remove_ucpu;
    my ($system_cpu) = $create_scpu + $remove_scpu;
    my (%summary);
    $summary{'volumes'} = scalar @dirs;
    $summary{'dirs_per_volume'} = int $dirs;
    $summary{'total_dirs'} = $dirs * $summary{'volumes'};
    $summary{'files_per_dir'} = int $files_per_dir;
    $summary{'total_files'} = $files_per_dir * $summary{'total_dirs'};
    $summary{'blocksize'} = int $blocksize;
    $summary{'blocks_per_file'} = int $block_count;
    $summary{'filesize'} = $blocksize * $block_count;
    $summary{'data_size'} = $summary{'filesize'} * $summary{'total_files'};
    $extras{'summary'} = \%summary;
    report_results($data_start_time, $data_end_time, $data_elapsed_time,
		   $user_cpu, $system_cpu, \%extras);
}

run_workload(\&runit, $processes);

File: lib/clusterbuster/pod_files/files.py
#!/usr/bin/env python3

import os
import time
import subprocess
import mmap
import shutil

from clusterbuster_pod_client import clusterbuster_pod_client


class files_client(clusterbuster_pod_client):
    """
    Small files test for clusterbuster
    """

    def __init__(self):
        try:
            super().__init__()
            if len(self._args) > 6:
                self.dir_list = self._args[6:]
            else:
                self.dir_list = ['/var/opt/clusterbuster']
            self.dirs = self._toSize(self._args[0])
            self.files_per_dir = self._toSize(self._args[1])
            self.blocksize = self._toSize(self._args[2])
            self.block_count = self._toSize(self._args[3])
            self._set_processes(int(self._args[4]))
            self.o_direct = self._toBool(self._args[5])
            self.flags = 0
            if self.o_direct:
                self.flags = os.O_DIRECT
        except Exception as err:
            self._abort(f"Init failed! {err} {' '.join(self._args)}")

    def remdir(self, dirname: str, oktofail: bool = False):
        try:
            os.rmdir(dirname)
        except Exception as err:
            if oktofail:
                shutil.rmtree(dirname, ignore_errors=True)
            else:
                raise err

    def makethem(self, pid: int):
        buf = mmap.mmap(-1, self.blocksize)
        buf.write(b'a' * self.blocksize)
        ops = 0
        files_created = 0
        for bdir in self.dir_list:
            direc = f"{bdir}/{self.localid}"
            os.makedirs(direc)
            ops = ops + 2
            for subdir in range(self.dirs):
                dirname = f"{direc}/{subdir}"
                os.mkdir(dirname)
                ops = ops + 1
                for fileidx in range(self.files_per_dir):
                    filename = f"{dirname}/{fileidx}"
                    try:
                        fd = os.open(filename, self.flags | os.O_WRONLY | os.O_CREAT)
                    except Exception as exc:
                        raise Exception(f"Create failed on {filename} (file {files_created}): {exc}") from None
                    ops = ops + 1
                    files_created = files_created + 1
                    for block in range(self.block_count):
                        try:
                            answer = os.write(fd, buf)
                            if answer != self.blocksize:
                                raise os.IOError(f"Incomplete write to {filename} (file {files_created}): {answer} bytes, expect {self.blocksize}")
                            ops = ops + 1
                        except IOError as exc:
                            raise Exception(f"Write failed to {filename} (file {files_created}): {exc}") from None
                    try:
                        os.close(fd)
                    except IOError as exc:
                        raise Exception(f"Unable to close {filename} (file {files_created}): {exc}") from None
        return ops

    def readthem(self, pid: int, oktofail: bool = False):
        dbuf = ''
        ops = 0
        for bdir in self.dir_list:
            direc = f"{bdir}/{self.localid}"
            ops = ops + 2
            for subdir in range(self.dirs):
                dirname = f"{direc}/{subdir}"
                ops = ops + 1
                for fileidx in range(self.files_per_dir):
                    filename = f"{dirname}/{fileidx}"
                    try:
                        if self.o_direct and self.block_count > 0 and self.blocksize > 0:
                            fd = os.open(filename, self.flags | os.O_RDONLY)
                            ops = ops + 1
                            with mmap.mmap(fd, 0, prot=mmap.PROT_READ) as mm:
                                for block in range(self.block_count):
                                    tmp = mm.read(self.blocksize)
                                    dbuf += str(tmp[-1:])
                                    ops = ops + 1
                            os.close(fd)
                        else:
                            with open(filename) as file:
                                ops = ops + 1
                                for block in range(self.block_count):
                                    file.read(self.blocksize)
                                    ops = ops + 1
                    except Exception as exc:
                        if not oktofail:
                            raise exc
        return ops

    def removethem(self, pid: int, oktofail: bool = False):
        ops = 0
        for bdir in self.dir_list:
            direc = f"{bdir}/{self.localid}"
            if oktofail and not self._isdir(direc):
                continue
            for subdir in range(self.dirs):
                dirname = f"{direc}/{subdir}"
                if oktofail and not self._isdir(subdir):
                    continue
                for fileidx in range(self.files_per_dir):
                    filename = f"{dirname}/{fileidx}"
                    if oktofail and not self._isfile(filename):
                        continue
                    os.unlink(filename)
                    ops = ops + 1
                self.remdir(dirname, oktofail)
                ops = ops + 1
            self.remdir(direc, oktofail)
            ops = ops + 1
        return ops

    def run_one_operation(self, op_name0: str, op_name1: str, op_name2: str, op_func, pid: int, data_start_time: float):
        self._sync_to_controller(self._idname([pid, f"start {op_name2}"]))
        self._drop_cache()
        ucpu, scpu = self._cputimes()
        op_start_time = self._adjusted_time() - data_start_time
        ops = op_func(pid)
        op_end_time_0 = self._adjusted_time() - data_start_time
        self._drop_cache()
        op_end_time = self._adjusted_time() - data_start_time
        op_elapsed_time = op_end_time - op_start_time
        op_elapsed_time_0 = op_end_time_0 - op_start_time
        ucpu, scpu = self._cputimes(ucpu, scpu)
        answer = {
            'operation_elapsed_time': op_elapsed_time,
            'user_cpu_time': ucpu,
            'system_cpu_time': scpu,
            'cpu_time': ucpu + scpu,
            'cpu_utilization': (ucpu + scpu) / op_elapsed_time,
            'operation_start': op_start_time,
            'operation_end': op_end_time,
            'operations': ops,
            'operations_per_second': ops / op_elapsed_time
            }
        if op_name2 == 'read':
            answer['total_files'] = self.files_per_dir * self.dirs * len(self.dir_list)
            answer['self.block_count'] = self.block_count
            answer['size'] = self.blocksize
            answer['data_size'] = self.blocksize * self.block_count * answer['total_files']
            answer['data_rate'] = answer['data_size'] / op_elapsed_time_0
        self._timestamp(f'{op_name1} files...')
        self._sync_to_controller(self._idname([pid, f'end {op_name2}']))
        return answer

    def runit(self, process: int):
        for tree in self.dir_list:
            self._cleanup_tree(tree)
        self.localid = self._idname(separator='-')
        self.removethem(process, True)
        data_start_time = self._adjusted_time()

        subprocess.run('sync')
        answer_create = self.run_one_operation('Creating', 'Created', 'create', self.makethem, process, data_start_time)
        self._timestamp("Sleeping for 60 seconds")
        time.sleep(60)
        self._timestamp('Back from sleep')
        answer_read = self.run_one_operation('Reading', 'Read', 'read', self.readthem, process, data_start_time)
        self._timestamp("Sleeping for 60 seconds")
        time.sleep(60)
        self._timestamp('Back from sleep')
        answer_remove = self.run_one_operation('Removing', 'Remove', 'remove', self.removethem, process, data_start_time)
        create_et = answer_create['operation_end'] - answer_create['operation_start']
        # read_et = answer_read['operation_end'] - answer_read['operation_start']
        remove_et = answer_remove['operation_end'] - answer_remove['operation_start']
        data_start_time = answer_create['operation_start']
        data_end_time = answer_remove['operation_end']
        user_cpu = answer_create['user_cpu_time'] + answer_remove['user_cpu_time']
        system_cpu = answer_create['system_cpu_time'] + answer_remove['system_cpu_time']
        extras = {
            'summary': {
                'volumes': len(self.dir_list),
                'dirs_per_volume': self.dirs,
                'total_dirs': self.dirs * len(self.dir_list),
                'self.files_per_dir': self.files_per_dir,
                'total_files': self.files_per_dir * self.dirs * len(self.dir_list),
                'self.blocksize': self.blocksize,
                'blocks_per_file': self.block_count,
                'filesize': self.blocksize * self.block_count,
                'data_size': self.blocksize * self.block_count * self.files_per_dir * self.dirs * len(self.dir_list)
                },
            'create': answer_create,
            'read': answer_read,
            'remove': answer_remove
            }
        self._report_results(data_start_time, data_end_time, create_et + remove_et, user_cpu, system_cpu, extras)


files_client().run_workload()
